python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        client: temporalio.Client,
+        *,
+        # Uses default in client if not present (or errors if no default)
+        task_queue: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activity(
+        # This must be a callable. Can be a lambda.
+        activity,
+        *,
+        # Set the name override. If not set, uses the @activity.run decorator
+        # name. If that's not set, uses the function/method name or class name
+        # if it has __call__. If it's a lambda or nested function, a name must
+        # be provided (or it can be decorated of course).
+        name: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activities(
+        # This must be an object or list or dictionary. If an object, this only
+        # uses methods with @activity.run decorators. Dictionaries and lists
+        # don't require the decorator. If a dictionary, the key is the name.
+        # This is just a helper and is the equivalent of just calling
+        # register_activity with all applicable functions.
+        activities,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
+        pass
+```
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* Internally, activities will run in the same event loop `run()` is called on
+* By default will fail already-registered activities
+* Should we accept a set of activities to register in `__init__`?
+* Does exporting `temporalio.Worker` directly cause any packaging issues if we want to separate the client as its own
+  published package?
+
+### Activity Definition/Registration
+
+```python
+from temporalio import activity
+
+# This adds activity details if they are present. Maybe we should have a
+# temporalio.LoggerAdapter that works for activities _and_ workflows?
+my_activity_logger = activity.LoggerAdapter(logging.getLogger(__name__))
+
+
+def simple_activity(arg1, arg2):
+    # Uses contextvars internally, which means it'll fail in another context but
+    # we will provide the ability to set this for testing situations
+    attempts = activity.info().attempts
+    # This logger has an adapter to include custom details but still named after
+    # this module which is a common pattern
+    my_activity_logger.debug("Attempt %s", attempts)
+    # This logger, on the activity module, uses the name of the activity module.
+    # So that's why we offer either the adapter or the logger.
+    activity.logger().debug("Attempt %s", attempts)
+
+
+@activity.run(name="MyActivity")
```

Comment:
From below:

> * We use `@activity.run` from `temporalio.activity` instead of just `@activity` from because:
>   * We want `activity` to be a module because we want it usually imported as `from temporalio import activity` so
>     top-level calls like `activity.info()` read well
>   * When we get to workflows, we will do similar with `@workflow.run` vs `@workflow.signal` vs `@workflow.query`, etc
>     and `workflow.info()`

It will match `@workflow.run` a bit better. If we ever get activity signals, we'll be happy we did this, heh.

Reply:
Sure, but, my suggestion applies equally to `workflow.run`. I think the suggested alternatives fit it just as well

Reply:
Agreed, but I don't like them as much either. The others didn't fit as well:

* `@activity.def` - triggered the highlighter for the reserved `def`
* `@activity.define` - sounds like you're doing something instead of just marking it
* `@activity.definition`/`@activity.defn` - there's no `@workflow` equivalent because I want to mark where the entry is, not the entire workflow
* `@activity.fun`, `@activity.function` - "function" doesn't apply to all callables

But I too am not sure I like `@activity.run`. The main thing I'm thinking about from a workflow POV is marking the entrypoint of the workflow (not to be confused with marking the entire workflow). Then I just wanted to have an activity equivalent. Other options:

* `@activity.entry` - This doesn't fit activity as well as `@workflow.entry` because of course where you call the activity is its entrypoint
* `@activity.exec` - This actually works as well as `@workflow.exec`
* `@activity.call` - More the action than the marker (though can say the same for "run" and "exec", feels uglier here)

I have scoured the thesaurus. I am open to other options.

> Bikeshed territory, but an important one.

Very important IMO

Reply:
`handler` ? `.code`?

I dunno, `defn` might be best. For a workflow, sure, it's the "entry" and not the "entire workflow", but I don't really feel that's an important distinction here. The intent is pretty clear there.

Reply:
I can get behind `@activity.defn`. One of my secret ideas was to have a concept of a mutually exclusive `@workflow.run` and `@workflow.start` and if the latter was used, it'd just assume an infinite loop and be event based, heh. I think I can still get there with `@workflow.defn` though.